### PR TITLE
Hcx Refactors

### PR DESCRIPTION
**Pull request by khavinshankar**
## Proposed Changes

- Added authorization based on roles for hcx.claim and hcx.policy
- Added authentication for hcx listener apis
- Added suggestions made in #1219 

 
@coronasafe/code-reviewers

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete

### CODE REVIEW

**Comment by vigneshhari** [comment]
Fix conflicts

File: config/authentication.py [hunk 1]
```diff
@@ -208,6 +209,56 @@ def get_user(self, validated_token):
...
+            raise AuthenticationFailed("Invalid signature")
+        except Exception as e:
+            raise InvalidToken({"detail": str(e)})
+
+        user = self.get_user()
+
+        return user, payload
+
+    def get_jwt_token(self, token):
+        return token.replace("Bearer", "").replace(" ", "")
```
Comment (sainak) [inline config/authentication.py:234]:
```suggestion
        return token.replace("Bearer ", "", 1)
```
and raise TokenError if the result is empty string

File: config/authentication.py [hunk 2]
```diff
@@ -208,6 +209,56 @@ def get_user(self, validated_token):
...
 
 
+class HCXAuthentication(JWTAuthentication):
+    def authenticate_header(self, request):
+        return "Bearer"
+
+    def authenticate(self, request):
+        jwt_token = request.META.get("HTTP_AUTHORIZATION")
+        if jwt_token is None:
+            return None
```
Comment (sainak) [inline config/authentication.py:219]: this can be removed, in favor of token error in `get_jwt_token`

File: config/authentication.py [hunk 3]
```diff
@@ -208,6 +209,56 @@ def get_user(self, validated_token):
         return user
 
 
+class HCXAuthentication(JWTAuthentication):
```
Comment (sainak) [inline config/authentication.py:212]: add a simple schema definition class for swagger [ref](https://drf-spectacular.readthedocs.io/en/latest/customization.html?highlight=OpenApiAuthenticationExtension#specify-authentication-with-openapiauthenticationextension)

similar to
https://github.com/coronasafe/care/blob/0143084df7e994e2a45624a8ce684880782f1bf2/config/authentication.py#L272-L289

**Review by sainak — CHANGES_REQUESTED** [review]

**Review by Ashesh3 — APPROVED** [review]

**Review by sainak — APPROVED** [review]

**Comment by khavinshankar** [comment]
Closing this PR as this will be added as a part of HCX Plug PR

**Pull request closed without merging**